fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMerger.java

```diff
@@ -108,27 +112,70 @@ private CompletableFuture<Boolean> handleSuccess(final IndexDeferredMaintenanceC
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug(mergerLogMessage("IndexMerge: Success", mergeControl));
         }
-        // after a successful merge, reset the time quota. It is unlikely to be applicable for the next merge.
+        // after a successful merge, reset the time quota and document count. They are unlikely to be applicable for the next merge.
         timeQuotaMillis = 0;
+        repartitionDocumentCount = 0;
         // Here: no errors, stop the iteration unless has more
-        final boolean hasMore = mergeControl.getMergesFound() > mergeControl.getMergesTried();
+        final boolean hasMore =
+                shouldGiveRepartitionSecondChance() ||
+                        mergeControl.getMergesFound() > mergeControl.getMergesTried();
         return  hasMore ? AsyncUtil.READY_TRUE : AsyncUtil.READY_FALSE;
     }
 
     private CompletableFuture<Boolean> handleFailure(final IndexDeferredMaintenanceControl mergeControl, Throwable e) {
         final FDBException ex = IndexingBase.findException(e, FDBException.class);
+        final IndexDeferredMaintenanceControl.LastStep lastStep = mergeControl.getLastStep();
         if (!IndexingBase.shouldLessenWork(ex)) {
             giveUpMerging(mergeControl, e);
         }
-        // Here: this exception might be resolved by reducing the number of merges or forcing shorter intervals between auto-commits
-        if (mergeControl.getMergesTried() < 2) {
-            handleSingleMergeFailure(mergeControl, e);
-        } else {
-            handleMultiMergeFailure(mergeControl, e);
+        switch (lastStep) {
+            case REBALANCE:
+                // Here: this exception might be resolved by reducing the number of documents to move during rebalancing
+                handlePotentialRebalanceFailure(mergeControl, e);
```

Comment:
nit: perhaps there's no need for `Potential` in this method name, it is known that there is a failure, and it was because of rebalance.

Reply:
✅ 
(Good point!)